Arrow array

taylor12805 opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Support Array data type for inserting and selecting data in Arrow, Parquet and ORC formats

Detailed description / Documentation draft:

Support array data type in Parquet, Arrow and ORC storage formats by implement arrow array reader and writer in ArrowColumnToCHColumn.cpp and CHColumnToArrowColumn.cpp. I added tests for arrays with different types, including nullable and for nested arrays for these formats.
This PR is the result of the work of issue: https://github.com/ClickHouse/ClickHouse/issues/21080

taylor12805 committed:
update

taylor12805 committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into arrow_array

taylor12805 committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into arrow_array

taylor12805 committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into arrow_array

taylor12805 committed:
parquet data type support array format read and write

filimonov commented:
Please check / fix build failures. 

```
2021-03-17 09:50:53 /build/src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:650:33: error: variable 'list_nested_type' may be uninitialized when used here [-Werror,-Wconditional-uninitialized]
2021-03-17 09:50:53                         switch (list_nested_type)
2021-03-17 09:50:53                                 ^~~~~~~~~~~~~~~~
2021-03-17 09:50:53 /build/src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:564:13: note: variable 'list_nested_type' is declared here
2021-03-17 09:50:53             arrow::Type::type list_nested_type;
2021-03-17 09:50:53             ^
```

taylor12805 committed:
Merge branch 'master' into arrow_array

filimonov commented:
@taylor12805 do you need help?

taylor12805 commented:
> @taylor12805 do you need help?

I think i can handle it. Have no time last week, will continue this weekend

taylor12805 committed:
fix list_nested_type uninitialized compile error

taylor12805 commented:
Hi, filimonov, I have fixed variable not initialized bug. 
I tried to add array nullable support in the latest commit, but it ran into "Memory Address not mapped to object" error, which i had no idea why it happens. Could you please take a look at it? Thanks a lot
Test case:
`
CREATE TABLE test.event_table(`buvid` String, `extended_field` Array(Nullable(String))) ENGINE=File(Parquet)

insert into test.event_table(buvid, extended_field) values ('123', ['124', '4124']), ('124', ['124']),('125', ['124', '24444']),('126', ['124', '24444']),('127', ['124', '24444']),('128', ['41'])
`
Stack Trace is:

`
2021.04.11 22:20:26.882479 [ 12121 ] {} <Fatal> BaseDaemon: Address: 0x7f0cc6600000 Access: read. Address not mapped to object.

2021.04.11 22:20:26.882565 [ 12121 ] {} <Fatal> BaseDaemon: Stack trace: 0xea355d9 0xa6e42bb 0xa6de1a7 0xa6ddb26 0xa6dec96 0xa6e162f 0xa738b4a 0xa69d229 0xa652a68 0xa9078be 0xa8ec4ec 0xa90c4d0 0xa90c5cd 0xa6429d1 0xa643a9e 0xd15be6b 0xd15c378 0xd26a346 0xd267470 0x7f0dc072d609 0x7f0dc0654293

2021.04.11 22:20:26.884419 [ 12121 ] {} <Fatal> BaseDaemon: 3.1. inlined from /usr/lib/gcc/x86_64-linux-gnu/10/include/emmintrin.h:703: _mm_loadu_si128(long long vector[2] const*)

2021.04.11 22:20:26.884513 [ 12121 ] {} <Fatal> BaseDaemon: 3.2. inlined from /home/chou/Documents/ClickHouse/base/glibc-compatibility/memcpy/memcpy.h:194: inline_memcpy

2021.04.11 22:20:26.884550 [ 12121 ] {} <Fatal> BaseDaemon: 3. /home/chou/Documents/ClickHouse/base/glibc-compatibility/memcpy/memcpy.cpp:5: memcpy @ 0xea355d9 in /home/chou/Documents/ClickHouse_Build/programs/clickhouse

2021.04.11 22:20:26.937564 [ 12121 ] {} <Fatal> BaseDaemon: 4.1. inlined from /home/chou/Documents/ClickHouse/contrib/arrow/cpp/src/arrow/buffer_builder.h:137: arrow::BufferBuilder::UnsafeAppend(void const*, long)

2021.04.11 22:20:26.937641 [ 12121 ] {} <Fatal> BaseDaemon: 4.2. inlined from /home/chou/Documents/ClickHouse/contrib/arrow/cpp/src/arrow/buffer_builder.h:115: arrow::BufferBuilder::Append(void const*, long)

2021.04.11 22:20:26.937684 [ 12121 ] {} <Fatal> BaseDaemon: 4.3. inlined from /home/chou/Documents/ClickHouse/contrib/arrow/cpp/src/arrow/buffer_builder.h:211: arrow::TypedBufferBuilder<unsigned char, void>::Append(unsigned char const*, long)

2021.04.11 22:20:26.937740 [ 12121 ] {} <Fatal> BaseDaemon: 4. /home/chou/Documents/ClickHouse/contrib/arrow/cpp/src/arrow/array/builder_binary.h:64: arrow::BaseBinaryBuilder<arrow::BinaryType>::Append(unsigned char const*, int) @ 0xa6e42bb in /home/chou/Documents/ClickHouse_Build/programs/clickhouse

2021.04.11 22:20:26.966860 [ 12121 ] {} <Fatal> BaseDaemon: 5.1. inlined from /home/chou/Documents/ClickHouse/contrib/arrow/cpp/src/arrow/status.h:373: arrow::Status::operator=(arrow::Status&&)

2021.04.11 22:20:26.966938 [ 12121 ] {} <Fatal> BaseDaemon: 5.2. inlined from /home/chou/Documents/ClickHouse/src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:197: fillArrowArrayWithStringColumnData<DB::ColumnString>

2021.04.11 22:20:26.966981 [ 12121 ] {} <Fatal> BaseDaemon: 5. /home/chou/Documents/ClickHouse/src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:262: DB::CHColumnToArrowColumn::fillArrowArray(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, COW<DB::IColumn>::immutable_ptr<DB::IColumn>&, std::__1::shared_ptr<DB::IDataType const> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::shared_ptr<arrow::Array>, DB::PODArray<char8_t, 4096ul, Allocator<false, false>, 15ul, 16ul> const*, arrow::ArrayBuilder*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) @ 0xa6de1a7 in /home/chou/Documents/ClickHouse_Build/programs/clickhouse

2021.04.11 22:20:26.999244 [ 12121 ] {} <Fatal> BaseDaemon: 6.1. inlined from /home/chou/Documents/ClickHouse/contrib/libcxx/include/string:2218: std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::~basic_string()

2021.04.11 22:20:26.999322 [ 12121 ] {} <Fatal> BaseDaemon: 6. /home/chou/Documents/ClickHouse/src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:167: DB::fillArrowArrayWithArrayColumnData(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, COW<DB::IColumn>::immutable_ptr<DB::IColumn>&, std::__1::shared_ptr<DB::IDataType const> const&, std::__1::shared_ptr<arrow::Array>, DB::PODArray<char8_t, 4096ul, Allocator<false, false>, 15ul, 16ul> const*, arrow::ArrayBuilder*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) @ 0xa6ddb26 in /home/chou/Documents/ClickHouse_Build/programs/clickhouse

2021.04.11 22:20:27.028725 [ 12121 ] {} <Fatal> BaseDaemon: 7. /home/chou/Documents/ClickHouse/src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:278: DB::CHColumnToArrowColumn::fillArrowArray(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, COW<DB::IColumn>::immutable_ptr<DB::IColumn>&, std::__1::shared_ptr<DB::IDataType const> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::shared_ptr<arrow::Array>, DB::PODArray<char8_t, 4096ul, Allocator<false, false>, 15ul, 16ul> const*, arrow::ArrayBuilder*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) @ 0xa6dec96 in /home/chou/Documents/ClickHouse_Build/programs/clickhouse

2021.04.11 22:20:27.051740 [ 12121 ] {} <Fatal> BaseDaemon: 8.1. inlined from /home/chou/Documents/ClickHouse/contrib/libcxx/include/string:2218: std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::~basic_string()

2021.04.11 22:20:27.051796 [ 12121 ] {} <Fatal> BaseDaemon: 8. /home/chou/Documents/ClickHouse/src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:438: DB::CHColumnToArrowColumn::chChunkToArrowTable(std::__1::shared_ptr<arrow::Table>&, DB::Block const&, DB::Chunk const&, unsigned long, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) @ 0xa6e162f in /home/chou/Documents/ClickHouse_Build/programs/clickhouse

2021.04.11 22:20:27.077388 [ 12121 ] {} <Fatal> BaseDaemon: 9.1. inlined from /home/chou/Documents/ClickHouse/contrib/libcxx/include/string:2218: std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::~basic_string()

2021.04.11 22:20:27.077449 [ 12121 ] {} <Fatal> BaseDaemon: 9. /home/chou/Documents/ClickHouse/src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp:39: DB::ParquetBlockOutputFormat::consume(DB::Chunk) @ 0xa738b4a in /home/chou/Documents/ClickHouse_Build/programs/clickhouse
`

taylor12805 committed:
add array nullable datatype support

Avogar commented on src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:
```diff
@@ -91,20 +94,95 @@ namespace DB
+            arrow::Status components_status;
+
+            const auto & offsets = internal_column.getOffsets();
+            const IColumn & data = internal_column.getData();
```
The problem is here. If nested type is nullable then data will be ColumnNullable, but we want nested column. Try this:
```c++
const IColumn & data = is_column_array_nullable ? static_cast<const ColumnNullable &>(internal_column.getData()).getNestedColumn() : internal_column.getData();

```

Avogar commented on src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:
```diff
@@ -91,20 +94,95 @@ namespace DB
+
+            for (size_t idx = 0, size = internal_column.size(); idx < size; ++idx)
+            {
+                if (null_bytemap && (*null_bytemap)[idx])
```
Seems like this check is wrong. In case of array null_bytemap refers to the nested column, we don't have `Nullable(Array)`

Avogar commented on src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:
```diff
@@ -91,20 +94,95 @@ namespace DB
+                        }
+                        else
+                        {
+                            PaddedPODArray<UInt8> * array_nested_null_bytemap = new PaddedPODArray<UInt8>();
```
Why do we need new/delete here? Why don't just create PaddedPODArray on stack?

Avogar commented:
@taylor12805 I found the problem, please fix it.

Avogar commented:
Oops, I didn't want to close it, it was a misclick, sorry.

Avogar commented:
Oops, I didn't want to close it, it was a misclick, sorry.

Avogar commented on src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:
```diff
@@ -173,9 +245,71 @@ namespace DB
+        ColumnPtr & nested_column,
+        const std::shared_ptr<const IDataType> & column_nested_type,
+        const String column_nested_type_name,
+        std::shared_ptr<arrow::Array> arrow_array,
```
I see you removed using `arrow_array` from all places except fillArrowArrayWithDecimalColumnData, why? And I see that `Array(Decimal)` is not supported by now, will you add it? If you need help, ask me.

Avogar commented on src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:
```diff
@@ -249,6 +384,28 @@ namespace DB
+                        column.name, arrow::list(arrow_type_it->second), is_column_array_nullable));
+                } else
+                {
+                    throw Exception{"The type \"" + column_array_nested_type_name + "\" of a array column \"" + column.name + "\""
```
By now using `Array(Decimal)` throws here.

Avogar commented on src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:
```diff
@@ -91,20 +94,95 @@ namespace DB
+                    offset_length = offsets[idx] - offset_start;
+                    if (offset_length > 0)
+                    {
+                        auto cut_data = data.cut(offset_start, offset_length);
```
Here we do unnecessary coping of data. We can avoid it by adding two arguments `size_t row_num, size_t size` to fillArrowArray* functions and insert only range column[row_num, row_num + size].

taylor12805 replied:
That's a good idea, I will implement it this weekend

taylor12805 replied:
And fix other problems mentioned above. Thanks a lot for reviewing the code

Avogar commented on src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:
```diff
@@ -75,6 +94,35 @@ namespace DB
     }
 
+    template <typename NumericType, typename VectorType = ColumnVector<NumericType>>
+    static void fillColumnWithArrayNumericData(std::shared_ptr<arrow::ChunkedArray> & arrow_column, MutableColumnPtr & internal_column)
```
Maybe we can combine this code with fillColumnWithNumericData to reduce code duplication? ListArray has methods like values() and offsets(), so we can firstly get offsets and then insert values without iterating over arrays.

Avogar commented:
@taylor12805 How is it going? Do you need help?

faisalhasnain commented:
I am very keen to see when it will get merged?

Avogar commented:
> I am very keen to see when it will get merged?

let's give the author a week to respond, and then we'll see

taylor12805 commented:
> > I am very keen to see when it will get merged?
> 
> let's give the author a week to respond, and then we'll see

Sorry, I am working on the implementation. Have only weekend time to do it, so it is not going quite smoothly. Have a vacation from 5.1 to 5.5, I try to finish it druing this time

buyology commented:
@taylor12805 — any update on this? Would you accept PRs into your branch to help out? (Not saying I'm the best suited to make such a patch, probably someone more versed in CH could finish it off more quickly.) 

We would really love for this to go in soon.

alexey-milovidov commented:
@buyology You can also create another PR with changes on top of this PR.

taylor12805 commented:
add array nullable datatype support in ArrowColumnToCHColumn, gonna add array(decimal) support this week

taylor12805 committed:
add array nullable datatype support in ArrowColumnToCHColumn and fix select from array datatype misplaced bug

Commit:
Refactor, support all types in array, add nested arrays support, more tests

Avogar commented:
I've decided to finish this PR by myself, because this feature is important and many people are waiting for it. I refactored the code, made it more pretty and readable, added missing types support in array, removed coping from some places, added nested arrays support, added more tests. BTW, @taylor12805 thank you so much for you work.

Commit:
Some fixes

Commit:
Try to fix tests

Commit:
Change tests

Commit:
Try fix tests again

taylor12805 commented:
> I've decided to finish this PR by myself, because this feature is important and many people are waiting for it. I refactored the code, made it more pretty and readable, added missing types support in array, removed coping from some places, added nested arrays support, added more tests. BTW, @taylor12805 thank you so much for you work.

Sorry couldn't make more support. Was quite a newbie in ClickHouse development when I started this PR. 
I will try to come up with a complete work next time I contribute. @Avogar

Commit:
Remove debug output

Avogar commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into arrow_array

Commit:
Final fixes

Commit:
Remove fillArrowArray from .h file

Commit:
Remove include_indices

Commit:
Return include_indices, fix arrays and add more tests
